Implement the rest of CEL

Flaque opened the pull request:
# Overview
This PR adds support for the rest of CEL.

We @ sfcompute.com are interested in using CEL for ourselves, but would like a reasonable typescript implementation to exist.

So this PR hooks up the conformance tests from [the CEL spec](https://github.com/google/cel-spec) into a test framework, and then iteratively makes all of them pass. 

It adds functionally every missing feature and resolves a lot of the edge cases from the existing implementation. 

Apologies for the large PR! 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added tests (if possible)
- [x] I have updated the readme (if necessary)
- [x] I have updated the demo (if necessary)

Flaque committed:
implement all the literals

Flaque committed:
add .all

Flaque committed:
add exists

Flaque committed:
add exists_one macro

Flaque committed:
add filter

Flaque committed:
add map

Flaque committed:
add larger, more real-world integration test suite

Flaque committed:
add duration & chained expressions

Flaque committed:
add contains, endsWith, trim, and split

Flaque committed:
implement math functions

Flaque committed:
add types

Flaque committed:
update demo and types

Flaque committed:
add conformance tests

Flaque committed:
comparison 32% => 72%; string 58% => 64%; list from 41% to 61%

Flaque committed:
pass even more basic tests

Flaque committed:
more string compliance tests fixed

Flaque committed:
even more conformance tests pass

Flaque committed:
even more conformance tests passed

Flaque committed:
add parser tests

Flaque committed:
fix even more conformance tests

Flaque committed:
even more conformance tests

Flaque committed:
more conformance tests passing

Flaque committed:
fix more tests

Flaque committed:
add more passing tests

Flaque committed:
add yet more conformance tests

Flaque committed:
pass more tests

Flaque committed:
pass more conformance tests

Flaque committed:
more conformance tests pass

Flaque committed:
pass even more tests

Flaque committed:
fix more conformance tests

Flaque committed:
more conformance tests

Flaque committed:
nuke the custom parser and use @bufbuild/cel-spec

Flaque committed:
pass more conformance tests

Flaque committed:
add buf loader

Flaque committed:
fix more conformance tests

Flaque committed:
more conformance tests

Flaque committed:
fix more tests

Flaque committed:
even more tests passing

Flaque committed:
add more datils

Flaque committed:
add more details

Flaque committed:
update demo

Flaque committed:
Resolve merge conflicts in PR #62

- Merged optional chaining and quoted identifier support with existing collection macros
- Combined CST definitions from both branches
- Integrated parser changes for better identifier handling
- Resolved visitor conflicts while preserving all functionality

ChromeGG commented:
Whoa, what a wonderful PR!
 I'll review & run it tomorrow morning (in 10h). Please fix the conflicts with main branch 
Amazing job :)

Flaque committed:
Resolve merge conflicts from ChromeGG/cel-js main

- Merged optional chaining and quoted identifier support
- Kept collection macros functionality
- Resolved conflicts in cst-definitions.d.ts, parser.ts, and visitor.ts

Flaque committed:
fix tests

ChromeGG committed:
fix some lint errors

ChromeGG committed:
remove duplicated tests, fix demo.ts

ChromeGG committed:
format example file

ChromeGG commented:
Please take care of these issues so I can proceed with the review:

## 1. Eslint Issues
There are 575 eslint problems (538 errors, 37 warnings). While not all of them are significant, a large portion is, such as:
```ts
const result: any = {} 
// can be
const result: Record<string, unknown> = {}

// no IF statements in the tests, etc.
// no unused variables
```

## 2. Code Formatting

Please run prettier on all the files to ensure they align with the repository conventions.

## 3. Comments

Please address the comments I left.

---

Also the `visitor.ts` becomes a monster (over 6k LOC). IDK if it's an issue but it can be hard to maintain. But I have no idea how to refactor it 😅 But it's just nice-to-have fix, not a blocker

SksOp commented:
Hey i just saw this PR, I was about to open a few issues without knowing this exists. @Flaque are you still working on this? I also need this in my project and would like to know the progress and if you need some help.

ChromeGG commented:
@SksOp I think there is no progress on it. You can fix the issue I listed above. Happy to merge it once it's resolved
